### PR TITLE
replace yapf with ruff

### PR DESCRIPTION
**Pull request by remyleone**
Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: .github/workflows/format.yml [hunk 1]
```diff
@@ -47,7 +47,7 @@ jobs:
       run: |
         isort --diff --check --profile black -l 88 -m 3 \
             sky/skylet/providers/ibm/
-    - name: Running isort for yapf formatted files
+    - name: Running isort for ruff formatted files
```
Comment (hemildesai) [inline .github/workflows/format.yml:50]: Ruff has isort capabilities as well so this step can ideally be merged with the above

Comment (remyleone) [reply .github/workflows/format.yml:50]: Do you know how to configure it in the ruff settings?

**Review by Michaelvll — COMMENTED** [review]
Thanks for submitting the PR @remyleone! I was trying this PR by manually making some changes of the code. I added a new line to L155 of `sky/provision/aws/instance.py`
https://github.com/skypilot-org/skypilot/blob/9743aa0c7f86a352715b13dd18ee780239f190f6/sky/provision/aws/instance.py#L155
 to make it: 
```python
    _merge_tag_specs(tag_specs,
user_tag_specs)
```
It seems the original formatter can change it back, but the ruff's based formatter fail to format it. Do you have an idea why that happens?

**Comment by remyleone** [comment]
I've tried running `ruff format .` from the root directory and it seems to work to change this specific line. 🤔

**Comment by remyleone** [comment]
I would suggest iterating with other PR. Given ruff speed, it does not seems to be a major issue to reformat gradually the whole codebase.

**Comment by Michaelvll** [comment]
Thanks for the reply @remyleone! It seems when I ran `./format.sh` it did not change the file, but when I run `./format.sh --all`, it will take effect. However, `ruff` seems changing all our quoting from single to double quotes, which is not align with our code style. Could you help see if there is a way to keep the behavior of the formatter the same as our previous `yapf` formatter using the google style?

**Comment by remyleone** [comment]
It could be changed by adding:

```
[tool.ruff.format]
quote-style = "single"
```

in pyproject.toml

That being said, it seems that there is places in the code that are using [double quote](https://github.com/skypilot-org/skypilot/blob/master/examples/distributed_ray_train/train.py#L60) instead of single quote. Could we merge first this PR and then have additional PR to fix issue and iterating?

**Comment by Michaelvll** [comment]
> It could be changed by adding:
> 
> ```
> [tool.ruff.format]
> quote-style = "single"
> ```
> 
> in pyproject.toml
> 
> That being said, it seems that there is places in the code that are using [double quote](https://github.com/skypilot-org/skypilot/blob/master/examples/distributed_ray_train/train.py#L60) instead of single quote. Could we merge first this PR and then have additional PR to fix issue and iterating?

Hi @remyleone, thanks for the suggestion! Would you like to update the PR accordingly?
If the ruff is going to update the quotes used in the examples, it should be good to have those changes in this PR as well. We probably would like to merge this PR after it can guarantee that a developer running the formatter will not get significant changes to the existing code, and fix any incorrect formatting that would have been fixed by our previous `yapf`. : )

File: .github/workflows/format.yml (hunk 1 above)
Comment (hemildesai) [reply .github/workflows/format.yml:49]: Sorry for the late reply. Here's the documentation to add it as a pylint rule - https://docs.astral.sh/ruff/rules/#isort-i with example config - https://docs.astral.sh/ruff/faq/#how-does-ruffs-import-sorting-compare-to-isort

**Pull request closed without merging**